File: Data Structures & Algorithms/binary-search/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        def binarySearch(arr: List[int], left: int, right: int):
            mid = int((left + right) / 2)

            if left > right:
                return - 1
            
            if arr[mid] == target:
                return mid
            elif arr[mid] < target:
                return binarySearch(arr, mid + 1, right)
            else:
                return binarySearch(arr, left, mid - 1)
        
        left = 0
        right = len(nums) - 1
        return binarySearch(nums, left, right)
            

File: Data Structures & Algorithms/search-insert-position/submission-3.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (target > nums[mid]) {
                // end of array
                if (mid == nums.length - 1) {
                    return mid + 1;
                }
                if (target < nums[mid + 1]) {
                    return mid + 1;
                }
                left = mid + 1;

            } else if (target < nums[mid]) {
                // start of array
                if (mid == 0)  {
                    return mid;
                }
                if (target > nums[mid - 1]) {
                    return mid;
                }
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return nums.length + 1;
    }
}